did-doc-diff/src/index.ts

```diff
@@ -0,0 +1,217 @@
+}) {
+  if (await validate_sig(tick)) {
+    return tick.diffs
+  } else {
+    return undefined
+  }
+}
+
+export async function tick_from_diffs(
+  diffs: { [index: string]: any },
+  key: crypto.EcdsaKeypair,
+) {
+  // This function dose not validate the diffs remember to do that first
+  const tids = Object.keys(diffs).sort()
+  if (tids.length < 1) {
+    return undefined // ToDo: handel the case where aic has no diffs
+  }
+
+  const tick = {
+    tid: TID.next().formatted(),
+    diffs: diffs,
+    id:
+      'did:aic:' +
+      (await pid(uint8arrays.fromString(JSON.stringify(diffs[tids[0]])))),
+    key: key.did(),
+    sig: '',
+  }
+  const signed_tick = sign(tick, key)
+  return signed_tick
+}
+
+function authorise_key(doc: { [index: string]: any }, didkey: string) {
+  const valid = Boolean(
+    (doc['adx/account_keys'] && doc['adx/account_keys'].includes(didkey)) ||
+      (doc['adx/recovery_keys'] && doc['adx/recovery_keys'].includes(didkey)),
+  )
+  console.log('authorise_key:', valid)
+  return true
+}
+
+export async function diffs_to_did_doc(diffs: { [index: string]: any }) {
+  if (typeof diffs !== 'object') {
+    return undefined
+  }
+  const tids = Object.keys(diffs).sort()
+  if (tids.length < 1) {
+    return undefined
+  }
+  let last = tids[0]
+  let doc = diffs[last]
+
+  // add the id feild
+  doc['id'] =
+    'did:aic:' +
+    (await pid(uint8arrays.fromString(JSON.stringify(diffs[last]))))
+
+  for (const tid of tids.slice(1)) {
+    const diff = diffs[tid]
+    if (!(await validate_sig(diff))) {
+      console.log('Warning ignoreing: bad sig')
+      continue // not valid ignore it
+    }
+    if (!authorise_key(doc, diff.key)) {
+      console.log('Warning ignoreing: bad key')
+      continue // not valid ignore it
+    }
+
+    const { prev, patches } = diff
+    if (prev !== last) {
+      console.log('Warning ignoreing: out of order')
+      continue
+    }
+    doc = patch(doc, patches)
+    last = tid
+  }
+  return doc
+}
+
+const S32_CHAR = '234567abcdefghijklmnopqrstuvwxyz'
+const B32_CHAR = 'abcdefghijklmnopqrstuvwxyz234567'
+
+function s32encode(data: Uint8Array): string {
+  // this is gross so many alocatons and funtion calls :(
+  const base32 = uint8arrays.toString(data, 'base32')
+  const sort_order_invariant_base32 = base32.split('').map((c) => {
+    return S32_CHAR[B32_CHAR.indexOf(c)]
+  })
+  return sort_order_invariant_base32.join('')
+}
+
+async function pid(data: Uint8Array) {
+  // return the pid120 a 24 char pid
+  const hash = s32encode((await sha256.digest(data)).digest)
+  let twos = 0
+  for (twos = 0; twos < 32; twos++) {
+    if (hash[twos] !== '2') {
+      break
+    }
+  }
+  const head = S32_CHAR[31 - twos]
+  const tail = hash.slice(twos, twos + 15) // 23 for pid120; 15 for pid80 // Are we ok with 80 bits?
+  return head + tail
+}
+
+function patch(doc: { [index: string]: any }, patches: Array<any>) {
+  for (const p of patches) {
+    const [op, path, value] = p
```

Comment:
are there any constraints we should put on values?

and do we allow more complicated patches that aren't just `put`/`del`? such as `append` for adding something to a list?

Reply:
now if you want to append you put the whole list again.
I only allow navigating down objects 
```json
{"a": [{"b": "c"}]}
```
```json
["put", ["a", 0, "b"], "not c"]
```
will get mad you would need to do put the whole list
```json
[ "put", ["a"], [{"b": "not c"}] ]
```
we can loosen this restriction late but for now made the code a little simpler for the type checker